File: src/Budgeteer.App/Models/Setup/Budget/ListModel.cs
// -----------------------------------------------------------------------------------------------------------------------------------------
// <copyright file="ListModel.cs" company="Lukas Tietze">
// Copyright (c) Lukas Tietze. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------------------------------------------------------------------------

namespace Budgeteer.App.Models.Setup.Budget;

/// <summary>
/// Modelliert die Daten zum Auflisten von Budgets.
/// </summary>
public class ListModel : ModelBase
{
    /// <summary>
    /// Holt oder setzt Betrag des Budgets.
    /// </summary>
    public long Amount { get; set; }

    /// <summary>
    /// Holt oder setzt die Sammlung der untergeordneten Budgets.
    /// </summary>
    public List<ListModel> Children { get; set; } = [];

    /// <summary>
    /// Holt oder setzt die ID des übergeordneten Knotens.
    /// </summary>
    public int? ParentId { get; set; }
}

File: src/Budgeteer.App/Client/ts/Models/Setup/ListModel.ts
import { ModelBase } from '../ModelBase';

/**
 * Stellt ein eingestelltes Budgets dar.
 */
export class ListModel extends ModelBase {
  /**
   * Initialisiert eine neue Instanz der Klasse.
   */
  constructor(copy?: Partial<ListModel>) {
    super(copy);

    this.parentId = copy?.parentId ?? 0;
    this.children = copy?.children?.map((c) => new ListModel(c)) ?? [];
  }

  /**
   * Die ID der übergeordneten Kategorie.
   */
  public parentId: number;

  /**
   * Holt oder setzt die untergeordneten Budgets.
   */
  public children: ListModel[];
}
